src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php

```diff
@@ -149,22 +155,69 @@ public function dump(array $options = array())
             }
         }
 
-        $code = $this->startClass($options['class'], $options['base_class'], $options['namespace']);
+        $code =
+            $this->startClass($options['class'], $options['base_class']).
+            $this->addServices().
+            $this->addDefaultParametersMethod().
+            $this->endClass()
+        ;
 
-        if ($this->container->isCompiled()) {
-            $code .= $this->addFrozenConstructor();
-            $code .= $this->addFrozenCompile();
-            $code .= $this->addFrozenIsCompiled();
+        if ($this->asFiles) {
+            $fileStart = <<<EOF
+<?php
+
+use Symfony\Component\DependencyInjection\Argument\RewindableGenerator;
+
+// This file has been auto-generated by the Symfony Dependency Injection Component for internal use.
+
+EOF;
+            $files = array();
+            foreach ($this->generateServiceFiles() as $file => $c) {
+                $files[$file] = $fileStart.$c;
+            }
+            foreach ($this->generateProxyClasses() as $file => $c) {
+                $files[$file] = "<?php\n".$c;
+            }
+            $files['Container.php'] = $code;
+            $hash = ucfirst(strtr(ContainerBuilder::hash($files), '._', 'xx'));
+            $code = array();
+
+            foreach ($files as $file => $c) {
+                $code["Container{$hash}/{$file}"] = $c;
+            }
+            array_pop($code);
+            $code["Container{$hash}/Container.php"] = implode("\nclass Container{$hash}", explode("\nclass {$options['class']}", $files['Container.php'], 2));
+            $namespaceLine = $this->namespace ? "\nnamespace {$this->namespace};\n" : '';
+
+            $code[$options['class'].'.php'] = <<<EOF
+<?php
+{$namespaceLine}
+// This file has been auto-generated by the Symfony Dependency Injection Component for internal use.
+
+if (!class_exists(Container{$hash}::class, false)) {
+    require __DIR__.'/Container{$hash}/Container.php';
+}
+
+if (!class_exists({$options['class']}::class, false)) {
+    /*{$this->docStar}
+     * {$options['class']}.
+     *
+     * @final since Symfony 3.3
+     */
+    class {$options['class']} extends Container{$hash}
+    {
+    }
+}
+
+return new Container{$hash}();
```

Comment:
This is a very important piece added just now, permitted by this PR. In fixes an architectural issue with the dumped class. Here, instead of defining the container class straight away, we turn the file into a container factory. For BC, a class is created with the expected name. But under the hood, we generated a container class whose name contains a hash of the dumped code. This class, and all service factory files are put into a directory that is named after this hash also.

This way, there can be no collision between two container's cache. This makes it easy to deal with atomicity when writting files to disk, and makes it easy also to generate two different container side-by-side (this last aspect will be leveraged into a later PR to fix the cache:clear command.)

So, now on, this PR is not only a small but nice perf improvement. It's also a architectural improvement.